geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala

```diff
@@ -54,142 +256,231 @@ object SimpleFeatureRecordConverter {
+    override def convertToAttribute(value: java.lang.Byte): Integer = value.intValue()
   }
 
-  class MapToRecordField(name: String, valueType: AttributeFieldConverter[AnyRef, AnyRef])
-      extends AttributeFieldConverter[java.util.Map[AnyRef, AnyRef], java.util.Map[String, AnyRef]] {
+  class ShortFieldConverter(name: String)
+      extends AbstractFieldConverter[Integer, java.lang.Short](name, RecordFieldType.SHORT.getDataType) {
+    override def convertToRecord(value: Integer): java.lang.Short = value.shortValue()
+    override def convertToAttribute(value: java.lang.Short): Integer = value.intValue()
+  }
+
+  class BigIntFieldConverter(name: String)
+      extends AbstractFieldConverter[java.lang.Long, BigInteger](name, RecordFieldType.BIGINT.getDataType) {
+    override def convertToRecord(value: java.lang.Long): BigInteger = BigInteger.valueOf(value)
+    override def convertToAttribute(value: BigInteger): java.lang.Long = value.longValueExact()
+  }
+
+  class CharFieldConverter(name: String)
+      extends AbstractFieldConverter[String, java.lang.Character](name, RecordFieldType.CHAR.getDataType) {
+    override def convertToRecord(value: String): Character = value.charAt(0)
+    override def convertToAttribute(value: Character): String = value.toString
+  }
+
+  class ListFieldConverter(val name: String, sub: FieldConverter[AnyRef, AnyRef])
+      extends FieldConverter[java.util.List[AnyRef], Array[AnyRef]] {
+
     override val field: RecordField =
-      new RecordField(name, RecordFieldType.MAP.getMapDataType(valueType.field.getDataType))
-    override def toRecord(attribute: java.util.Map[AnyRef, AnyRef]): java.util.Map[String, AnyRef] = {
-      if (attribute == null) { null } else {
-        attribute.asScala.map { case (k, v) => k.toString -> valueType.toRecord(v) }.asJava
-      }
-    }
-    override def toAttribute(record: java.util.Map[String, AnyRef]): java.util.Map[AnyRef, AnyRef] = {
-      if (record == null) { null } else {
-        // TODO records do not support non-string map keys
-        record.asScala.map { case (k, v) => k -> valueType.toAttribute(v) }.asJava.asInstanceOf[java.util.Map[AnyRef, AnyRef]]
-      }
+      new RecordField(name, RecordFieldType.ARRAY.getArrayDataType(sub.field.getDataType))
+
+    override val descriptor: AttributeDescriptor =
+      new AttributeTypeBuilder()
+          .binding(classOf[java.util.List[AnyRef]])
+          .buildDescriptor(name)
+          .setListType(sub.descriptor.getType.getBinding)
+
+    override def convertToRecord(value: java.util.List[AnyRef]): Array[AnyRef] =
+      value.asScala.collect { case v if v != null => sub.convertToRecord(v) }.toArray
+
+    override def convertToAttribute(value: Array[AnyRef]): java.util.List[AnyRef] =
+      java.util.Arrays.asList(value.collect { case v if v != null => sub.convertToAttribute(v) }: _*)
+  }
+
+  // TODO records do not support non-string map keys
```

Comment:
What situation isn't covered here?

Reply:
our Map-type attributes are not fully supported, since we support any binding for the keys (e.g. you could have `Map[Int, Int]`). i think we should probably deprecate Map-type attributes since we have json fields now that are much more flexible...